crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,564 @@
+//! A buffered implementation that automatically groups native prices API
+//! requests into batches.
+
+use {
+    crate::price_estimation::{
+        native::{NativePriceEstimateResult, NativePriceEstimating},
+        PriceEstimationError,
+    },
+    anyhow::anyhow,
+    futures::{
+        channel::mpsc,
+        future::FutureExt as _,
+        stream::{self, FusedStream, Stream, StreamExt as _},
+    },
+    primitive_types::H160,
+    std::{
+        collections::{HashMap, HashSet},
+        future::Future,
+        num::NonZeroUsize,
+        sync::Arc,
+        time::Duration,
+    },
+    tokio::{sync::broadcast, task::JoinHandle},
+};
+
+/// Buffered configuration.
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to request.
+    ///
+    /// Specifying `None` means no limit on concurrency.
+    pub max_concurrent_requests: Option<NonZeroUsize>,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay to start counting after receiving the first request.
+    pub debouncing_time: Duration,
+    /// The timeout to wait for the result to be ready
+    pub result_ready_timeout: Duration,
+    /// Maximum capacity of the broadcast channel to store the native prices
+    /// results
+    pub broadcast_channel_capacity: usize,
```

Comment:
Any reason to not use `usize::max` for this one

Reply:
@fleupold Because if something goes south (one requested, aka, consumer is not reading the broadcast channel), then the elements from the broadcast won't be deleted until hit the limit, and if the limit is the `usize::MAX`, the memory can leak. If this scenario happens, the missing consumer will miss data (if the broadcast limit is 20 for example), but if it is not consuming the data in the first place...

Reply:
Sure, I just don't think we will ever change whatever value we deem appropriate here for a different use case and it's making the configuration object harder to understand. We can keep it but imo it shouldn't be something an engineer should think about when setting up a buffered price estimator (whatever you think is a reasonable value is likely good for most if not all use cases)

Reply:
> Small addition. [...]

Ok wow, even more reason I should never be forced to chose this value myself

Reply:
I'm fine with hardcoding the value eventually but this seems like the sort of thing which might require a bit of time to dial in to not over-allocate memory but also don't miss results unnecessarily. I'd say let's leave it as a config for now and hardcode it once we have found a reasonable lower bound in practice.

Reply:
I fully agree with @MartinquaXD